ddtrace/internal/_encoding.pyx

```diff
@@ -91,180 +91,223 @@ cdef class Packer(object):
-
-                    ret = pack_bytes(&self.pk, <char *>b"name", 4)
-                    if ret != 0: return ret
-                    ret = self._pack(o.name)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"error", 5)
-                    if ret != 0: return ret
-                    ret = self._pack(1 if o.error else 0)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"start", 5)
-                    if ret != 0: return ret
-                    ret = self._pack(o.start_ns)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"duration", 8)
-                    if ret != 0: return ret
-                    ret = self._pack(o.duration_ns)
-                    if ret != 0: return ret
-
-                    if has_span_type:
-                        ret = pack_bytes(&self.pk, <char *>b"type", 4)
-                        if ret != 0: return ret
-                        ret = self._pack(o.span_type)
-                        if ret != 0: return ret
-
-                    if has_meta:
-                        ret = pack_bytes(&self.pk, <char *>b"meta", 4)
-                        if ret != 0: return ret
-                        ret = self._pack(o.meta)
-                        if ret != 0: return ret
-
-                    if has_metrics:
-                        ret = pack_bytes(&self.pk, <char *>b"metrics", 7)
-                        if ret != 0: return ret
-                        ret = self._pack(o.metrics)
-                        if ret != 0: return ret
-            else:
-                PyErr_Format(TypeError, b"can not serialize '%.200s' object", Py_TYPE(o).tp_name)
+        if metrics is None:
+            ret = msgpack_pack_nil(&self.pk)
+
+        if PyDict_CheckExact(metrics):
+            d = <dict> metrics
+            L = len(d)
+            if L > ITEM_LIMIT:
+                raise ValueError("dict is too large")
+
+            ret = msgpack_pack_map(&self.pk, L)
+            if ret == 0:
+                for k, v in d.items():
+                    ret = self._pack_text(k)
+                    if ret != 0: break
+                    ret = self._pack_number(v)
+                    if ret != 0: break
             return ret
 
+        raise TypeError("Unhandled metrics type: %r" % type(metrics))
+
+    cdef inline int _pack_span(self, object span):
+        cdef int ret
+        cdef Py_ssize_t L
+        cdef int has_span_type
+        cdef int has_meta
+        cdef int has_metrics
+
+        has_span_type = <bint> (span.span_type is not None)
+        has_meta = <bint> (len(span.meta) > 0)
+        has_metrics = <bint> (len(span.metrics) > 0)
+
+        L = 9 + has_span_type + has_meta + has_metrics
+
+        ret = msgpack_pack_map(&self.pk, L)
+
+        if ret == 0:
+            ret = pack_bytes(&self.pk, <char *> b"trace_id", 8)
```

Comment:
aside: I wonder if we can write this data directly into the `self.pk.buf`... ?

it might be a bad idea, using `pack_bytes` which uses the `msgpack_pack_raw`/`msgpack_pack_raw_body` is probably still the safest option we have.

Reply:
(note, this is not a change worth making in this PR, it is a musing, do not make this change here)

Reply:
> it might be a bad idea, using `pack_bytes` which uses the `msgpack_pack_raw`/`msgpack_pack_raw_body` is probably still the safest option we have.

I also think this is the best thing to do. I'd expect those two API calls to be optimised already.

Reply:
thanks for the gut check 👍🏻